_posts/2019-02-06-Age-Targeting-Part2.md

```diff
@@ -0,0 +1,254 @@
+```
+
+
+**Feature creation**<br/>
+The question to keep in mind is: how do people of different age use their smart devices?
+The more difference can be seen between different classes in a feature, the better it is.
+Only the ones which contribute to improve the prediction are kept in the final model.
+
+* **Time distribution**<br/>
+We calculate the hourly and weekly usage distribution, namely the fraction of the total time spent on apps split by hour or by day of the week.
+In the plot we can see the average of the hourly fraction calculated for each classes.
+The trends follow the expectations: the younger tend to spend more time on the apps late in the evening, while the older like to be active earlier in the morning.
+<p align="center"> <img src="../images/Age-Targeting/parallel_hour.png"> </p>
+
+* **The apps**<br/>
+We collect the used apps over the last month for each device.
+This feature is directly related to the different kind of usages for a smart devices.
+As we will see later in the post, it has the strongest relation to the age.
+To be passed as a feature in the machine learning model, we transform the list of used apps to a [bag of words](https://en.wikipedia.org/wiki/Bag-of-words_model) via the CountVectorizer class in Spark.
+In this way every apps constitutes a single feature.
+
+```python
+from pyspark.ml.feature import CountVectorizer, Normalizer
+
+apps_count_vectorizer = CountVectorizer(inputCol=feature_col, 
+                                        outputCol=count_vectorizer_col, 
+                                        minDF=min_df)
+apps_normalizer = Normalizer(p=2.0, 
+                             inputCol=count_vectorizer_col, 
+                             outputCol=normalizer_col)
+```
+The minDF parameter of the CountVectorizer class is used to cut off apps visited by few devices to decrease the noise in the dataset and the chance to overfit.
+
+* **Make and model**<br/>
+Both make and model can be expressed as categorical features and treated via [One Hot Encoding](https://en.wikipedia.org/wiki/One-hot).
+Due to the high cardinality of the models we map the low frequent occurrences into a single index to reduce the sparsity of the data and reduce the overfitting.
+We know that certain models can be more appealing for younger people while others could have different functionality more appreciated in the adult phase (e.g. a large screen).
+
+```python
+from pyspark.ml.feature import OneHotEncoderEstimator
+
+make_model_ohe = OneHotEncoderEstimator(inputCols=[make_col, model_col], 
+                                        outputCols=[make_col_out, model_col_out])
+```
+
+* **The home country**<br/>
+The home country of the device is preprocessed as a categorical feature via One Hot Encoding in the same way of the make and model.
+It is a proxy for many different quantities such as economy wealthy, age distribution, etc...
+
+* **Number of apps**<br/>
+Number of distinct apps used by the device.
```

Comment:
This sentence no verb. I suggest to rename the title to "Number of distinct apps" and remove that sentence.